src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -87,115 +109,119 @@ EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_sh
+				bundled_assembly_index >= 1) {
+			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, entry_name.get (), apk_name);
+			if (register_debug_symbols_for_assembly (entry_name, (bundled_assemblies) [bundled_assembly_index - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
 				continue;
 		}
 
 #if !defined(NET6)
-		if (utils.ends_with (file_name, ".config") && bundled_assemblies != nullptr) {
-			char *assembly_name = strdup (basename (file_name));
+		if (utils.ends_with (entry_name, ".config") && !bundled_assemblies.empty ()) {
+			char *assembly_name = strdup (basename (entry_name.get ()));
 			// Remove '.config' suffix
 			*strrchr (assembly_name, '.') = '\0';
 
-			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
 			mono_register_config_for_assembly (assembly_name, (const char*)map_info.area);
 
 			continue;
 		}
 #endif // ndef NET6
 
-		if (!utils.ends_with (file_name, ".dll"))
+		if (!utils.ends_with (entry_name, ".dll"))
 			continue;
 
 		if (entry_is_overridden)
 			continue;
 
-		size_t alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(void*), bundled_assemblies_count + 1);
-		bundled_assemblies = reinterpret_cast<MonoBundledAssembly**> (utils.xrealloc (bundled_assemblies, alloc_size));
-		MonoBundledAssembly *cur = bundled_assemblies [bundled_assemblies_count] = reinterpret_cast<MonoBundledAssembly*> (utils.xcalloc (1, sizeof (MonoBundledAssembly)));
-		++bundled_assemblies_count;
+		assembly_count--;
+		MonoBundledAssembly *cur;
+		if (XA_UNLIKELY (assembly_count < 0 || (bundled_assemblies_slow_path && bundled_assemblies.size () <= bundled_assembly_index))) {
+			if (assembly_count == -1) {
+				log_warn (LOG_ASSEMBLY, "Number of assemblies stored at build time (%u) was incorrect, switching to slow bundling path.");
+			}
+			bundled_assemblies.emplace_back ();
+			cur = &bundled_assemblies.back ();
+		} else {
+			cur = &bundled_assemblies [bundled_assembly_index];
+		}
+		bundled_assembly_index++;
 
-		md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
-		cur->name = utils.monodroid_strdup_printf ("%s", strstr (file_name, prefix) + prefix_len);
+		md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
+		cur->name = utils.strdup_new (entry_name.get () + prefix_len, entry_name.length () - prefix_len);
 		cur->data = (const unsigned char*)map_info.area;
 
 		// MonoBundledAssembly::size is const?!
 		unsigned int *psize = (unsigned int*) &cur->size;
 		*psize = static_cast<unsigned int>(file_size);
 
-		if (utils.should_log (LOG_ASSEMBLY)) {
+		if (XA_UNLIKELY (utils.should_log (LOG_ASSEMBLY))) {
 			const char *p = (const char*) cur->data;
 
-			char header[9];
-			for (size_t j = 0; j < sizeof(header)-1; ++j)
+			std::array<char, 9> header;
```

Comment:
Sure, more C++20 use, but…really?  Not sure what `std::array` is buying us here.

Reply:
Likely "obvious" answer: https://coders-corner.net/2018/06/16/stdarray-vs-c-style-array

Good for `<algorithm>` integration, range checking in Debug builds, etc.

Not used *here*, but maybe eventually?

Reply:
Yes, it's that and it's for consistency around the code. Also, `std::array` is a zero-overhead abstraction over C style arrays so we don't sacrifice anything.